<div style="text-align: center; background-color: #0A6EBD; font-family: 'Trebuchet MS', Arial, sans-serif; color: white; padding: 30px; font-size: 27px; font-weight: bold; border-radius: 0 0 0 0; box-shadow: 0px 6px 8px rgba(0, 0, 0, 0.2);">
  Final Project - Programming For Data Science @ FIT-HCMUS, VNU-HCM
</div>

<div style="text-align: center; background-color: #5A96E3; font-family: 'Trebuchet MS', Arial, sans-serif; color: white; padding: 20px; font-size: 25px; font-weight: bold; border-radius: 0 0 0 0; box-shadow: 0px 6px 8px rgba(0, 0, 0, 0.2);">
  Data exploring
</div>

## 1. Import library

In [1]:
import numpy as np
import pandas as pd

## 2. Data exploring

### 2.1. Read raw data from file

In [2]:
raw_supermarket_df = pd.read_csv('../data/raw/supermarket_sales.csv',
                                  index_col='Invoice ID')
raw_supermarket_df.head()

Branch       City Customer type  Gender            Product line  \
Invoice ID                                                                    
750-67-8428      A     Yangon        Member  Female       Health and beauty   
226-31-3081      C  Naypyitaw        Normal  Female  Electronic accessories   
631-41-3108      A     Yangon        Normal    Male      Home and lifestyle   
123-19-1176      A     Yangon        Member    Male       Health and beauty   
373-73-7910      A     Yangon        Normal    Male       Sports and travel   

             Unit price  Quantity   Tax 5%     Total       Date   Time  \
Invoice ID                                                               
750-67-8428       74.69         7  26.1415  548.9715   1/5/2019  13:08   
226-31-3081       15.28         5   3.8200   80.2200   3/8/2019  10:29   
631-41-3108       46.33         7  16.2155  340.5255   3/3/2019  13:23   
123-19-1176       58.22         8  23.2880  489.0480  1/27/2019  20:33   
373-73-7910       86.31         7  30.2085  634.3785   2/8/2019  10:37   

                 Payment    cogs  gross margin percentage  gross income  \
Invoice ID                                                                
750-67-8428      Ewallet  522.83                 4.761905       26.1415   
226-31-3081         Cash   76.40                 4.761905        3.8200   
631-41-3108  Credit card  324.31                 4.761905       16.2155   
123-19-1176      Ewallet  465.76                 4.761905       23.2880   
373-73-7910      Ewallet  604.17                 4.761905       30.2085   

             Rating  
Invoice ID           
750-67-8428     9.1  
226-31-3081     9.6  
631-41-3108     7.4  
123-19-1176     8.4  
373-73-7910     5.3

### 2.2. How many rows and how many columns?

In [3]:
shape = raw_supermarket_df.shape
print(f'Data have {shape[0]} rows and {shape[1]} columns')

Data have 1000 rows and 16 columns


### 2.3. What is the meaning of each row?

Each row represents a single supermarket bill.

### 2.4. Are there duplicated rows?

In [4]:
num_duplicated_rows = raw_supermarket_df.duplicated().sum()
print(f'Number of duplicated rows: {num_duplicated_rows}')

Number of duplicated rows: 0


### 2.5. What is the meaning of each column?

Based on [Super Market Sales Data](https://www.kaggle.com/datasets/arunjangir245/super-market-sales/data)

| ID | Field            | Description                                                   |
|----|------------------|---------------------------------------------------------------|
| 1  | Invoice ID       | Unique identifier for each transaction                        |
| 2  | Branch           | Three-to-five character hashtag indicating the branch         |
| 3  | City             | Location of the branch                                        |
| 4  | Customer Type    | Indicates if the customer is regular or new                   |
| 5  | Gender           | Gender of the customer                                        |
| 6  | Product Line     | Category/type of purchased product                            |
| 7  | Unit Price       | Price of a single product unit                                |
| 8  | Quantity         | Number of units purchased                                     |
| 9  | Tax 5%           | Amount of 5% tax applied to the transaction                   |
| 10 | Total            | Total cost of the transaction, including tax                  |
| 11 | Date             | Transaction date                                              |
| 12 | Time             | Transaction time of day                                       |
| 13 | Payment          | Payment method used (e.g., credit card, cash)                 |
| 14 | COGS             | Direct costs related to producing/purchasing the products sold|
| 15 | Gross Margin %   | Profit margin percentage for the transaction                  |
| 16 | Gross Income     | Total profit earned from the transaction                      |
| 17 | Rating           | Customer satisfaction rating or feedback on the transaction   |


### 2.6. What is the current data type of each column? Are there columns having inappropriate data types?

In [5]:
raw_supermarket_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1000 entries, 750-67-8428 to 849-09-3807
Data columns (total 16 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Branch                   1000 non-null   object 
 1   City                     1000 non-null   object 
 2   Customer type            1000 non-null   object 
 3   Gender                   1000 non-null   object 
 4   Product line             1000 non-null   object 
 5   Unit price               1000 non-null   float64
 6   Quantity                 1000 non-null   int64  
 7   Tax 5%                   1000 non-null   float64
 8   Total                    1000 non-null   float64
 9   Date                     1000 non-null   object 
 10  Time                     1000 non-null   object 
 11  Payment                  1000 non-null   object 
 12  cogs                     1000 non-null   float64
 13  gross margin percentage  1000 non-null   float64
 14  gross income

In [6]:
raw_supermarket_df['Date'].head()

Invoice ID
750-67-8428     1/5/2019
226-31-3081     3/8/2019
631-41-3108     3/3/2019
123-19-1176    1/27/2019
373-73-7910     2/8/2019
Name: Date, dtype: object

The Date format is currently in m/d/YYYY, and our intention is to convert it to dd/mm/YYYY format. We will combine these values into a `Datetime` column while removing the `Date` and `Time` columns. The resulting `Datetime` column should be the datetime datatype.

In [7]:
raw_supermarket_df['Datetime'] = raw_supermarket_df['Date'] + ' ' +raw_supermarket_df['Time']
raw_supermarket_df['Datetime'] = pd.to_datetime(raw_supermarket_df['Datetime'], format='%m/%d/%Y %H:%M')
raw_supermarket_df = raw_supermarket_df.drop(columns=['Date', 'Time'])

In [8]:
# Check again
raw_supermarket_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1000 entries, 750-67-8428 to 849-09-3807
Data columns (total 15 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   Branch                   1000 non-null   object        
 1   City                     1000 non-null   object        
 2   Customer type            1000 non-null   object        
 3   Gender                   1000 non-null   object        
 4   Product line             1000 non-null   object        
 5   Unit price               1000 non-null   float64       
 6   Quantity                 1000 non-null   int64         
 7   Tax 5%                   1000 non-null   float64       
 8   Total                    1000 non-null   float64       
 9   Payment                  1000 non-null   object        
 10  cogs                     1000 non-null   float64       
 11  gross margin percentage  1000 non-null   float64       
 12  gross income          

### 2.7. With each numerical column, how are values distributed?

In [9]:
def get_numeric_col_info(df):
    numeric_cols = df.select_dtypes(include=[np.number]).columns
    titles = ["missing_ratio", "min", "lower_quartile", "median", "upper_quartile", "max"]

    data = {col: [df[col].isna().mean() * 100, *df[col].describe()[['min', '25%', '50%', '75%', 'max']].values]
            for col in numeric_cols}

    data['Titles'] = titles
    num_col_info_df = pd.DataFrame(data).set_index('Titles')
    return num_col_info_df

num_col_info_df = get_numeric_col_info(raw_supermarket_df)
num_col_info_df

Unit price  Quantity     Tax 5%        Total      cogs  \
Titles                                                                   
missing_ratio        0.000       0.0   0.000000     0.000000    0.0000   
min                 10.080       1.0   0.508500    10.678500   10.1700   
lower_quartile      32.875       3.0   5.924875   124.422375  118.4975   
median              55.230       5.0  12.088000   253.848000  241.7600   
upper_quartile      77.935       8.0  22.445250   471.350250  448.9050   
max                 99.960      10.0  49.650000  1042.650000  993.0000   

                gross margin percentage  gross income  Rating  
Titles                                                         
missing_ratio                  0.000000      0.000000     0.0  
min                            4.761905      0.508500     4.0  
lower_quartile                 4.761905      5.924875     5.5  
median                         4.761905     12.088000     7.0  
upper_quartile                 4.761905     22.445250     8.5  
max                            4.761905     49.650000    10.0

#### Remove meaningless rows and columns
Since the 'gross margin percentage' column only contains the value 4.761905, we have decided to remove it.

In [10]:
raw_supermarket_df = raw_supermarket_df.drop(columns=['gross margin percentage'])

### 2.8. With each categorical column, how are values distributed?

In [11]:
pd.set_option('display.max_colwidth', 300)
pd.set_option('display.max_columns', None)

def get_categorical_col_info(df):
    categorical_df = df.select_dtypes(include=['object', 'bool'])
    col_names = categorical_df.columns
    titles = ["missing_ratio", "num_values", "value_ratios"]
    values_list = []

    for col in col_names:
        size = len(categorical_df)
        missing_ratio = ((categorical_df[col].isnull().sum() / size) * 100).round(3)
        num_values = categorical_df[col].nunique()
        value_counts = categorical_df[col].value_counts(normalize=False).to_dict()
        values_list.append([missing_ratio, num_values, value_counts])

    data = {col_names[i]: values_list[i] for i in range(len(col_names))}
    data['Titles'] = titles
    cat_col_info_df = pd.DataFrame(data).set_index('Titles')
    return cat_col_info_df

cat_col_info_df = get_categorical_col_info(raw_supermarket_df)
cat_col_info_df

Branch  \
Titles                                          
missing_ratio                             0.0   
num_values                                  3   
value_ratios   {'A': 340, 'B': 332, 'C': 328}   

                                                             City  \
Titles                                                              
missing_ratio                                                 0.0   
num_values                                                      3   
value_ratios   {'Yangon': 340, 'Mandalay': 332, 'Naypyitaw': 328}   

                                Customer type                        Gender  \
Titles                                                                        
missing_ratio                             0.0                           0.0   
num_values                                  2                             2   
value_ratios   {'Member': 501, 'Normal': 499}  {'Female': 501, 'Male': 499}   

                                                                                                                                                                        Product line  \
Titles                                                                                                                                                                                 
missing_ratio                                                                                                                                                                    0.0   
num_values                                                                                                                                                                         6   
value_ratios   {'Fashion accessories': 178, 'Food and beverages': 174, 'Electronic accessories': 170, 'Sports and travel': 166, 'Home and lifestyle': 160, 'Health and beauty': 152}   

                                                         Payment  
Titles                                                            
missing_ratio                                                0.0  
num_values                                                     3  
value_ratios   {'Ewallet': 345, 'Cash': 344, 'Credit card': 311}

## 3. Is the collected data reasonable?                                         

### 3.1. Checking whether the columns `Branch` and `City` convey the same information or share a similar meaning within the dataset.

In [12]:
def checking_branch_city(df: pd.DataFrame) -> bool:
    """Checking for the reasonable of Branch and City column.

    Args:
        df (pd.DataFrame): _description_

    Returns:
        bool: _description_
    """
    unique_branch_city = raw_supermarket_df[['Branch', 'City']].drop_duplicates()
    
    if len(unique_branch_city) == 3:
        return True
    else:
        return False

In [13]:
checking_branch_city(raw_supermarket_df)

True

Since 'Branch' and 'City' contain similar information, we decided to remove 'Branch' from the dataset.

In [14]:
raw_supermarket_df = raw_supermarket_df.drop(columns=['Branch'])

### 3.2. Checking the reasonability of the `Tax 5%` column's values
The tax is calculated as follows:  
$$\text{Tax} = \text{Unit price} \times \text{Quantity} \times \frac{5}{100}$$ 

In [15]:
def checking_tax(df: pd.DataFrame, tolerance=1e-10) -> bool:
    """Checking for the reasonableness of the Tax column.

    Args:
        df (pd.DataFrame): DataFrame containing columns 'Tax 5%', 'Unit price', 'Quantity'.
        tolerance (float): Tolerance level for the acceptable difference. Default is 1e-10.

    Returns:
        bool: True if differences are within the tolerance, False otherwise.
    """
    calculated_tax = df['Unit price'] * df['Quantity'] * 0.05
    differences = abs(df['Tax 5%'] - calculated_tax)
    return (differences <= tolerance).all()

In [16]:
checking_tax(raw_supermarket_df)

True

### 3.3. Checking the reasonability of the `Total` column's values
The total price is calculated as follows:  
$$\text{Total} = \text{Unit price} \times \text{Quantity} + \text{Tax}$$
$$\implies \text{Total} = \text{Unit price} \times \text{Quantity} \times 1.05$$

In [17]:
def checking_total(df: pd.DataFrame, tolerance=1e-10) -> bool:
    """Checking for the reasonableness of the Total column.

    Args:
        df (pd.DataFrame): DataFrame containing columns 'Total', 'Unit price', 'Quantity'.
        tolerance (float): Tolerance level for the acceptable difference. Default is 1e-10.

    Returns:
        bool: True if differences are within the tolerance, False otherwise.
    """
    calculated_price = df['Unit price'] * df['Quantity'] * 1.05
    differences = abs(df['Total'] - calculated_price)
    return (differences <= tolerance).all()

In [18]:
checking_total(raw_supermarket_df)

True

## 4. Save processed data

In [19]:
save_name = 'supermarket_sales_processed.csv'
raw_supermarket_df.to_csv(f'../data/processed/{save_name}', index='Invoice ID')